jobhub.go

```diff
@@ -0,0 +1,190 @@
+	jobDependency   map[int][]int
+	startingJob     map[int]bool
+	recursionLevels map[int]int
+}
+
+type Job struct {
+	Name string
+	Path string
+
+	pipelineID int
+	id         int
+}
+
+type exitStatus struct {
+	runtime time.Duration
+	status  syscall.WaitStatus
+}
+
+func (j Job) String() string {
+	return fmt.Sprintf("ID (Name): %d (%s)", j.id, j.Name)
+}
+
+func NewPipeline() *Pipeline {
+	return &Pipeline{
+		Log:             logrus.StandardLogger(),
+		jobByID:         make(map[int]Job),
+		jobDependency:   make(map[int][]int),
+		startingJob:     make(map[int]bool),
+		recursionLevels: make(map[int]int),
+	}
+}
+
+func nextIDPipeline() int {
+	tempID := nextPipelineID
+	nextPipelineID++
+	return tempID
+}
+
+func (p *Pipeline) nextIDJob() int {
+	p.nextJobID++
+	return p.nextJobID
+}
+
+func (p *Pipeline) AddJob(job Job) Job {
+	for _, j := range p.jobContainer {
+		if j.id == job.id {
+			p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] | Job has already been added",
+				p.id, p.Name, job.id, job.Name)
+		}
+	}
+	if p.id == 0 {
+		p.id = nextIDPipeline()
+	}
+	job.pipelineID = p.id
+	job.id = p.nextIDJob()
+	p.jobContainer = append(p.jobContainer, job)
+	p.jobByID[job.id] = job
+	p.startingJob[job.id] = true
+	return job
+}
+
+func (p *Pipeline) AddJobDependency(job Job, deps ...Job) {
+	var tempContainer []Job
+	tempContainer = append(deps, job)
+	if p.id != 0 {
+		if p.jobContainer != nil {
+			for _, givenJob := range tempContainer {
+				if givenJob.id == p.id {
+					continue
+				} else {
+					p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] does not belong to this pipeline",
+						p.id, p.Name, givenJob.id, givenJob.Name)
+				}
+			}
+			for _, givenJob := range tempContainer {
+				for _, addedJob := range p.jobContainer {
+					if addedJob == givenJob {
+						continue
+					} else {
+						p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] has never been added",
+							p.id, p.Name, givenJob.id, givenJob.Name)
+					}
+				}
+			}
+			for _, d := range deps {
+				p.jobDependency[job.id] = append(p.jobDependency[job.id], d.id)
+				delete(p.startingJob, d.id)
+			}
+		} else {
+			p.Log.Panicf("Pipeline [%d][%s] | Job Container is empty",
+				p.id, p.Name)
+		}
+	} else {
+		p.Log.Panicf("Pipeline [%d][%s] | Pipeline not initialized",
+			p.id, p.Name)
+	}
+}
+
+func (p *Pipeline) resolveDependencyRecursion(jobID int, level int) {
+	if l := p.recursionLevels[jobID]; l >= level {
+		return
+	}
+	p.recursionLevels[jobID] = level
+	level++
+	for _, depID := range p.jobDependency[jobID] {
+		p.Log.Debugf("%s -> %s | recursion level: %d", p.jobByID[jobID], p.jobByID[depID], level)
+		p.resolveDependencyRecursion(depID, level)
+	}
+}
+
+func (p *Pipeline) resolveDependency() []int {
+	var queue []int
+	tempRL := make(map[int]int)
+	for startID, _ := range p.startingJob {
+		p.resolveDependencyRecursion(startID, 1)
+	}
+	for jID, l := range p.recursionLevels {
+		tempRL[jID] = l
+	}
+	for i := 0; i < len(tempRL); {
```

Comment:
This is like O(n^2) complexity. While it's not a problem for this scale, it will be a good excercidse to try `sort.SortSlice` interface ;-)

Please add a line: `TODO(amwolff) switch to sort.SortSlice` :)